src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComponentActivator.cs

```diff
@@ -0,0 +1,132 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
+{
+    public static class ComponentActivator
+    {
+        private static readonly Dictionary<string, IsolatedComponentLoadContext> s_AssemblyLoadContexts;
+        private static readonly Dictionary<IntPtr, Delegate> s_Delegates = new Dictionary<IntPtr, Delegate>();
+        private static readonly Func<IntPtr,string> s_marshalToString;
+
+        static ComponentActivator()
+        {
+#if PLATFORM_WINDOWS
+            s_AssemblyLoadContexts = new Dictionary<string, IsolatedComponentLoadContext>(StringComparer.InvariantCultureIgnoreCase);
+            s_marshalToString = nativeString => Marshal.PtrToStringUni(nativeString);
+#else
+            s_AssemblyLoadContexts = new Dictionary<string, IsolatedComponentLoadContext>(StringComparer.InvariantCulture);
+            s_marshalToString = nativeString => Marshal.PtrToStringUTF8(nativeString);
+#endif
+        }
+
+        /// <summary>
+        /// Native hosting entry point for creating a native delegate
+        /// </summary>
+        /// <param name="assemblyPathNative">Fully qualified path to assembly</param>
+        /// <param name="delegateTypeNative">Assembly qualified delegate type name</param>
+        /// <param name="typeNameNative">Assembly qualified type name</param>
+        /// <param name="methodNameNative">Public static method name compatible with delegateType</param>
+        /// <param name="functionHandle">Pointer where to store the function pointer result</param>
+        public static int CreateNativeDelegate(IntPtr assemblyPathNative,
+                                               IntPtr delegateTypeNative,
```

Comment:
I don't know interop well enough - so possibly an obvious question:
Why do we need to explicitly declare the type of the delegate?
Would it be possible to:
* Infer a delegate from the method itself (let's say it only has primitive type arguments)
* Use marshalling attributes on the method to specify the interop behavior

@AaronRobinsonMSFT would know better I assume.
I just find it a bit cumbersome, that in order to use this the managed side has to effectively declare its "native" interface as well as a separate thing (the delegate). I understand with the limited API surface we will never be able to load "arbitrary" method and work with it from the native code, but I was hoping that for some definition of "simple method" it would kind of work out of the box.

That said, if it's too complex to make something like that work, I'm fine with this as the first solution. I would just move the `delegateTypeNative` such that it can be made optional in the future.

Reply:
This dance with delegates is workaround for lack of function pointers in C#.

This will work better once we have this: https://github.com/dotnet/csharplang/blob/master/proposals/function-pointers.md

Reply:
This could be an instance where the term 'delegate' is being overloaded. It seems that @sdmaclea is taking this behavior is literally returning a type `Delegate`, but my interpretation has always been the term 'delegate' in this scenario isn't more about the concept rather than the type (i.e. function pointer).

Technically the `Delegate` type isn't needed because it can be inferred or built up. I would not make it optional. Either require it or remove it.

Reply:
I spent some time experimenting with this before I settled on this approach. I was trying to create the `DelegateType` from the `Type[]`, but the type info of a compiler generated delegate type didn't appear useful.

Type.GetMethod() needs to handle overloading. To get the one you want, there is an overload which takes an array of Types.

`CreateDelegate()` uses a `DelegateType` to chose the overload.

We can allow `delegateType = IntPtr.Zero`, we just need to chose a reasonable default.

My guess would be this could be a reasonable, easy to use default
```C#
delegate int DefaultDelegateType(string argsAsString)
```

Reply:
@jkotas and @sdmaclea What am I missing here with respect to building up a delegate type. It seems that given various forms of the static function name we should be able to determine which `Delegate` type to return, no?

Reply:
We need to explicitly specify the required call signature, because it is the only check of the call semantics.  `C++` will just cast and call the function ptr returned.

To specify the call signature:
1. Chose a specific required signature
2. Specify a Delegate type name (with an optional default)
3. Specify an array of argument and return types

In my opinion Delegate type name was cleanest.

As an implementation consideration.

To create a function pointer I need a `System.Delegate`.
To create a `System.Delegate` I need a `Type delegateType`
I do not know the syntax to specify the `Type delegateType` as a string so the Type.GetType() returns what I need. Unless the delegate type is generated by the compiler.

Reply:
> To create a function pointer I need a System.Delegate.

It is to create a function pointer using public APIs. The runtime can create function pointer without delegate. It is what `coreclr_create_delegate` does today.

Reply:
>I do not know the syntax to specify the Type delegateType as a string so the Type.GetType() returns what I need. Unless the delegate type is generated by the compiler.

@sdmaclea I don't really understand why that can't be done at runtime. I say that because of the following example:

``` csharp
Type type = Type.GetType(typeName, resolver, null);
MethodInfo mi = type.GetMethod("Foo");

// Inspect the MethodInfo and generate a Delegate Type on the fly
Type d = ...

return mi.CreateDelegate(d);
```

In this case there must be a single function and we can apply whatever policy about that export desired. If we don't do that then users must handle the case where a delegate must be defined prior to usage which seems like a weird requirement. Is dynamic code gen not an option here? Once that is in place we can use the function pointer proposal and the API remains clean.

Reply:
There is nothing actionable here...

Reply:
>There is nothing actionable here...

I don't know what that means. I am trying to understand why the approach above won't work? Is generating one on the fly not feasible for some reason?

Reply:
The way I read the comments from @jkotas and @AaronRobinsonMSFT is that there are 2 way to get a native function pointer to a method without having a pre-declared delegate type:
* runtime can do it - so we could expose a new QCall to do this and then use it here - basically the same functionality as `coreclr_get_delegate`
* using reflection APIs, build the delegate type dynamically and then use it to get the function pointer.

I think it's perfectly fine if we prescribe a certain way of how methods are going to be exposed to native code (calling convention, parameter marshalling and so on). Since all that requires is for the native app to declare a C/C++ function pointer with the right signature (knowing these rules). But there's no need for an explicit delegate type in the managed world.

If we do require the delegate type in the managed world:
* We get the advantage that other marshalling mechanisms and calling conventions can be used
* We require the developer to effectively specify the function declaration 3 times
  * The managed method signature
  * The managed delegate
  * The native function declaration

I don't know what is less work, if taking the code from `coreclr_get_delegate` or building it via reflection. I would not bother too much about perf in this case, we don't expect people to call this API many times (it should only act as an entry point).

Reply:
@AaronRobinsonMSFT You can use a generic Action or Func delegate. Something like "System.Action[[System.Int32,System.Runtime],[System.Int32,System.Runtime],[System.Int32,System.Runtime]],System.Runtime" would work, to avoid forcing customers to define many forms of delegate. Its not completely general, but its sufficient for most cases.

On Windows X86, there is an additional complication of calling convention, which may possibly cause problems as well.

Reply:
@davidwrighton Yay! Thank you. That was the missing context here. I didn't realize that was possible in this context. 

So basically the approach would be something like the following:

``` c++
CreateNativeDelegate(
  "fooassembly",
  "System.Action[[System.Int32,System.Runtime],[System.Int32,System.Runtime],[System.Int32,System.Runtime]],System.Runtime",
  "bartype",
  "StoreInts",
  &fp);
```

```csharp
class bartype
{
  void StoreInts(int a, int b);
  void StoreInts(int a, int b, int c); // Would be returned
}
```

This approach seems okay to me, but it is annoying the delegate type definition would need to be already contained in the assembly _or_ would need to be written out in that manner - which looks like IL.

I guess this alleviates my concern, but I do like @vitek-karas idea of creating a QCall. I will say if I want to consume a 3rd party assembly from native that doesn't export a `Delegate`, then I have to specify the awkward signature above which is annoying. I think this comes back to what scenario are we trying to unblock.

Reply:
> generic Action or Func delegate

Interop does not support generics delegates.

Reply:
> > There is nothing actionable here...
> 
> I don't know what that means

It means you posted while I was replying.

Reply:
>Interop does not support generics delegates.

Oh that's right. Hmm so I guess we are back to the users requiring the definition of a delegate then. That is a big boo. That does bring me back to my question about dynamically generating the `Delegate` type on the fly. What is the issue with that?

Reply:
Generating a new delegate type is a very heavyweight operation. Reflection.Emit is … not cheap, and its a concept which doesn't map well to aot scenarios that MonoVM may have.

Reply:
Notes:

* This is an internal API for to enable the `.NetHost.dll` to activate components.
* The proposed coreclr API exposes flexible delegate creation based on pre-exisiting compiled `delegateType`s. (These types are not required to be in the loaded assembly or its dependents)
* We are not required to expose the full set of functionality to the connecting coreclr/native host dll
* To write effective and useful NetHost component activation samples we will choose a finite set of method signatures. We can have predefined delegate types for these in Managed and `C`/`C++` headers.

Based on that set of observations.  I would recommend 
* leaving the coreclr API as is.  
* define a prototypical delegate. This seems like a logical match for the normal main entry point
```C#
public delegate int ComponentEntryPoint(string[] args);
```
* define NetHost `C`/`C++` API to use `ComponentEntryPoint`.
  * typedef/using define the corresponding function ptr  type
  * write sample code

Reply:
>To write effective and useful NetHost component activation samples we will choose a finite set of method signatures.

> define a prototypical delegate. This seems like a logical match for the normal main entry point

The above statements satisfy any concerns I have. Thanks!

Reply:
> This seems like a logical match for the normal main entry point
> `public delegate int ComponentEntryPoint(string[] args);`

I do not think that this signature is a good match for the scenario. Folks doing their own hosting typically want to exchange a bunch of pointers to start the hand-shake. We had API that took strings in desktop (`ExecuteInDefaultAppDomain`) and the typical usage was to format the pointers to strings on one side and then parse the strings back to pointers on the other side...

Reply:
If you would like to have one canonical entrypoint, it should take `void*`.

Reply:
@jkotas So 
```C#
public delegate IntPtr ComponentEntryPoint(IntPtr args);
```

Reply:
I would imagine the return should be an `int`. I was thinking about this signature and at first liked it a lot. However, after thinking about it more I can't really come up with a scenario where returning a pointer to something is going to be all that common. I absolutely think it should be one of the delegate types supported, but I think the canonical one make sense as an `int`.